whylog/teacher/__init__.py

```diff
@@ -0,0 +1,78 @@
+class Interval(object):
+    """
+    Represents interval in line.
+    """
+
+    def __init__(self, start_offset, length, abstract_line):
+        self._start_offset = start_offset
+        self._length = length
+        self._abstract_line = abstract_line
+
+    def get_start_offset(self):
+        return self._start_offset
+
+    def get_length(self):
+        return self._length
+
+    def get_location(self):
+        return self._abstract_line
+
+
+class Rule(object):
+    """
+    Represents rule entering by user.
+    """
+
+    def __index__(self, effect):
+        self._effect = effect
+        self._causes = set()
+
+    def add_cause(self, cause):
+        self._causes.add()
+
+    def remove_cause(self, cause):
+        self.causes.remove(cause)
```

Comment:
This code exists so that instead of:
```python
rule.causes.remove(cause)
```
we can call
```python
rule.remove_causes(cause)
```
as with adding causes. I don't think it would be so wrong if someone could access the `causes` set directly as a public field, which would let him do other set operations on it. This would remove some boilerplate code too and allow to, for instance, add many causes with just one line.